syntax = "proto3";

package LibreSVIP;

enum PluginCategory {
  INPUT = 0;
  OUTPUT = 1;
  MIDDLEWARE = 2;
}

enum ConversionMode {
  DIRECT = 0;
  SPLIT = 1;
  MERGE = 2;
}

enum ConflictPolicy {
  SKIP = 0;
  PROMPT = 1;
  OVERWRITE = 2;
  RENAME = 3;
}

message PluginInfo {
  string identifier = 1;
  string name = 2;
  string version = 3;
  string description = 4;
  string author = 5;
  string website = 6;
  string json_schema = 7;
  // additional fields for io plugins
  string file_format = 8;
  repeated string suffixes = 9;
  string icon_base64 = 10;
  string ui_json_schema = 11;
  string default_json_value = 12;
}

message PluginInfosRequest {
  PluginCategory category = 1;
  string language = 2;
}

message PluginInfosResponse {
  repeated PluginInfo values = 1;
}

message ConversionGroup {
  string group_id = 1;
  repeated string file_paths = 2;
}

message ConversionRequest {
  string input_format = 1;
  string output_format = 2;
  ConversionMode mode = 3;
  int32 max_track_count = 4;
  repeated ConversionGroup groups = 5;
  string input_options = 6;
  string output_options = 7;
  map<string, string> middleware_options = 8;
  string language = 9;
}

message SingleConversionResult {
  string group_id = 1;
  bool running = 2;
  bool completed = 3;
  string error_message = 4;
  repeated string warning_messages = 5;
}

message VersionRequest {
}

message VersionResponse {
  string version = 1;
}

message MoveFileRequest {
  string group_id = 1;
  string output_dir = 2;
  string stem = 3;
  string output_format = 4;
  ConflictPolicy conflict_policy = 5;
  bool force_overwrite = 6;
}

message MoveFileResponse {
  string group_id = 1;
  string output_path = 2;
  ConflictPolicy conflict_policy = 3;
  bool completed = 4;
  bool success = 5;
  string error_message = 6;
}

service Conversion {
  rpc PluginInfos(PluginInfosRequest) returns (PluginInfosResponse);
  rpc Convert(ConversionRequest) returns (stream SingleConversionResult);
  rpc Version(VersionRequest) returns (VersionResponse);
  rpc MoveFile(MoveFileRequest) returns (stream MoveFileResponse);
}
